feat(HACBS-2184): enable tekton task testing for apply-mapping

mmalina opened the pull request:
This adds a new PR workflow that installs a kind cluster, tekton, tkn and then runs tests (defined as pipelines) for tasks.

Documentation is added to CONTRIBUTING.md.

Two test pipelines for apply-mapping are included. One is a basic use case when there is no extra mapping file. The other is more elaborate with several components and a mapping file.

mmalina committed:
feat(HACBS-2184): enable tekton task testing for apply-mapping

This adds a new PR workflow that installs a kind cluster, tekton,
tkn and then runs tests (defined as pipelines) for tasks.

Documentation is added to CONTRIBUTING.md.

Two test pipelines for apply-mapping are included.
One is a basic use case when there is no extra mapping file.
The other is more elaborate with several components and
a mapping file.

Signed-off-by: Martin Malina <mmalina@redhat.com>

johnbieren commented on .github/scripts/test_tekton_tasks.sh:
```diff
@@ -0,0 +1,108 @@
+  echo "  Installing task"
+  kubectl apply -f $TASK_PATH
+
+  for TEST_PATH in ${TEST_PATHS[@]}
```
probably not necessary now, but it would be nice in the future to run the tests in parallel

mmalina replied:
Yeah, good point. I'd say typically you'll change only one or a few tasks at a time, so the overall runtime shouldn't be too long. Once you start running tests in parallel, you can no longer easily see all the output in the linear log - we would have to store it in files and saves those as artifacts. We can consider it in the future - if it starts to take too long.

mmalina commented on CONTRIBUTING.md:
```diff
@@ -82,3 +82,62 @@ Before a pull request can be merged:
+
+### Tekton Task Testing
+
+When a pull request is opened, for all the task version directories that are modified,
```
Thanks, fixed.

mmalina committed:
fix: add wait for pipeline to exist

Sometimes the test script failed if the pipeline
was not available immediately after creating it.

Signed-off-by: Martin Malina <mmalina@redhat.com>

johnbieren commented:
The code lgtm, but I see the apply-mapping test is taking 30 mins and counting since your last update to this PR, so something must be wrong

mmalina commented:
> The code lgtm, but I see the apply-mapping test is taking 30 mins and counting since your last update to this PR, so something must be wrong

Yeah, it failed to apply the task and the pipeline as well, so then it was waiting for the pipeline to appear forever (I added that loop because what I saw a few times happen was no error when applying the pipeline, but when I tried to run it, it would say it doesn't exist).

I'm not sure why this happened:
```
Installing task
Error from server (InternalError): error when creating "catalog/task/apply-mapping/0.3/apply-mapping.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": failed to call webhook: Post "[https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s](https://tekton-pipelines-webhook.tekton-pipelines.svc/defaulting?timeout=10s)": dial tcp 10.96.149.233:443: connect: connection refused
  Installing test pipeline: catalog/task/apply-mapping/0.3/tests/test-apply-mapping-empty.yaml
Error from server (InternalError): error when creating "catalog/task/apply-mapping/0.3/tests/test-apply-mapping-empty.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": failed to call webhook: Post "[https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s](https://tekton-pipelines-webhook.tekton-pipelines.svc/defaulting?timeout=10s)": dial tcp 10.96.149.233:443: connect: connection refused
```
The workflow is waiting for the tekton pods to be fully up, so not sure why it would blow up like this.

Anyway, I'll have to make it a bit more robust (and make it fail in a reasonable time instead of waiting forever). I will ask for review again once done.

mmalina committed:
fix: improve tekton installation

The mechanism for waiting for tekton pods didn't work properly.
Now we first wait for the pods to appear and then we wait
for them to be ready.

Signed-off-by: Martin Malina <mmalina@redhat.com>

mmalina committed:
feat: only trigger task tests workflow if tasks are changed

Signed-off-by: Martin Malina <mmalina@redhat.com>

mmalina commented:
@johnbieren please check the two new commits. It should be ready now.

johnbieren commented on .github/workflows/tekton_task_tests.yaml:
```diff
@@ -0,0 +1,55 @@
+      - name: Check cluster info
+        run: |
+          kubectl cluster-info
+      - name: Install Tekton
```
how long does this task run? shouldn't it have an overall timeout?

mmalina replied:
`kubectl apply` (1) is instant. Then waiting for the pods to appear (2) will typically take under 1 s. And finally it will typically take around 20 s for the pods to be ready (3).

So I don't think there is a significant risk that the while loop (2) will take forever. The biggest risk is that the pods will not be ready within 120 s and then the task testing will blow up in the later steps.

Maybe I should add another check of the pod readiness at the end of this step and just fail if they're still not ready? WDYT?

johnbieren replied:
sure, that sounds better. I am just a little hesitant to have while sleep loops with no timeout. Outages or whatever else may lead to things hanging so a timeout would be nice

mmalina replied:
Actually, the `tekton wait` will return a non-zero exit code if the timeout is exceeded, so we're fine there. Doing a timeout for a `while` loop in bash is [a little cumbersome](https://stackoverflow.com/questions/12321469/retry-a-bash-command-with-timeout) so instead I set up a step timeout of 10 minutes (default is 360). Please check.

mmalina committed:
feat: set a timeout for tekton install step

The default timeout is 4 hours. Let's kill
the step it's not done within 10 minutes.

This is really just a failsafe for the while
loop - if for whatever reason the tekton
pods never appear.

Signed-off-by: Martin Malina <mmalina@redhat.com>
